server/src/main/java/org/opensearch/search/aggregations/AggregatorTreeEvaluator.java

```diff
@@ -0,0 +1,93 @@
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Performs cost-benefit analysis on aggregator trees to optimize streaming decisions.
+ *
+ * <p>Evaluates whether streaming aggregations will be beneficial by analyzing the entire
+ * collector tree using {@link FlushModeResolver}. When streaming is determined to be
+ * inefficient, recreates the aggregator tree with traditional (non-streaming) aggregators.
+ * Decisions are cached to ensure consistency across concurrent segment processing.
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public final class AggregatorTreeEvaluator {
+
+    private AggregatorTreeEvaluator() {}
+
+    /**
+     * Analyzes collector tree and recreates it with optimal aggregator types.
+     *
+     * <p>Determines the appropriate {@link FlushMode} for the collector tree and recreates
+     * aggregators if streaming is not beneficial. Should be called after initial aggregator
+     * creation but before query execution.
+     *
+     * @param collector the root collector to analyze
+     * @param searchContext search context for caching and configuration
+     * @param aggProvider factory function to recreate aggregators when needed
+     * @return optimized collector (original if streaming, recreated if traditional)
+     * @throws IOException if aggregator recreation fails
+     */
+    public static Collector evaluateAndRecreateIfNeeded(
+        Collector collector,
+        SearchContext searchContext,
+        CheckedFunction<SearchContext, List<Aggregator>, IOException> aggProvider
+    ) throws IOException {
+        if (!searchContext.isStreamSearch()) {
+            return collector;
+        }
+
+        FlushMode flushMode = getFlushMode(collector, searchContext);
+
+        if (flushMode == FlushMode.PER_SEGMENT) {
+            return collector;
+        } else {
+            return MultiBucketCollector.wrap(aggProvider.apply(searchContext));
+        }
+    }
+
+    /**
+     * Resolves flush mode using cached decision or on-demand evaluation.
+     *
+     * @param collector the collector to evaluate
+     * @param searchContext search context for decision caching
+     * @return the resolved flush mode for this query
+     */
+    private static FlushMode getFlushMode(Collector collector, SearchContext searchContext) {
+        FlushMode cached = searchContext.getFlushMode();
+        if (cached != null) {
+            return cached;
+        }
+
+        long maxBucketCount = searchContext.getStreamingMaxBucketCount();
+        double minCardinalityRatio = searchContext.getStreamingMinCardinalityRatio();
+        long minBucketCount = searchContext.getStreamingMinBucketCount();
+        FlushMode mode = FlushModeResolver.resolve(collector, FlushMode.PER_SHARD, maxBucketCount, minCardinalityRatio, minBucketCount);
+
+        if (!searchContext.setFlushModeIfAbsent(mode)) {
+            FlushMode existingMode = searchContext.getFlushMode();
```

Comment:
we're checking existing flush mode after setting the flush mode?

Reply:
yeah - there could be race condition between different slices trying to create respective collectionManager and deciding. So, whichever wins, we just use that decision. This is to avoid conflicts between their decisions, we don't want to land into situation where different slices operate on different flushMode.